{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def clean_prediction_df(prediction_df):\n",
    "    '''\n",
    "    # Generate Allele Mapping\n",
    "    serotype_dict = json.load(open('serotype_dict.json'))\n",
    "    new_list = []\n",
    "    for serotype, alleles in serotype_dict.items():\n",
    "        for allele in alleles:\n",
    "            new_entry = {\n",
    "                'H source': allele.get('name'),\n",
    "                'H allele': '-'.join([serotype, str(allele.get('num'))])\n",
    "            }\n",
    "            new_list.append(new_entry)\n",
    "    allele_df = pd.DataFrame(new_list)\n",
    "    allele_df.to_csv('allele_mapping.csv', index_label=False)\n",
    "    '''\n",
    "    allele_df = pd.read_csv('allele_mapping.csv')\n",
    "    allele_df.head()\n",
    "\n",
    "    # Merge with allele data\n",
    "    prediction_df = prediction_df.merge(allele_df, left_on='H allele', right_on='H allele', how='left')\n",
    "    allele_df.columns = ['O allele', 'O source']\n",
    "    prediction_df = prediction_df.merge(allele_df, left_on='O allele', right_on='O allele', how='left')\n",
    "    prediction_df.head()\n",
    "\n",
    "    # Read ectyper output\n",
    "    # Data cleaning\n",
    "    prediction_df['genome name'] = prediction_df['genome name'].str.split('|', expand=True)[1]\n",
    "    prediction_df['predicted O'] = prediction_df['predicted O'].str.split('O', expand=True)[1]\n",
    "    prediction_df['predicted H'] = prediction_df['predicted H'].str.split('H', expand=True)[1]\n",
    "    prediction_df.index = prediction_df['genome name']\n",
    "    prediction_df = prediction_df[['predicted H','H source', 'predicted O', 'O source']]\n",
    "    prediction_df.head()\n",
    "\n",
    "    '''\n",
    "    # Get genome file paths as a list\n",
    "    import os\n",
    "    genome_dir = '/home/sam/Projects/MoreSerotype/temp/genomes'\n",
    "    genome_names = os.listdir(genome_dir)\n",
    "    genome_files = []\n",
    "    for genome_name in genome_names:\n",
    "        file = os.path.join(genome_dir, genome_name)\n",
    "        genome_files.append(file)\n",
    "\n",
    "    # Create serotype dataframe\n",
    "    from Bio import SeqIO\n",
    "    import re\n",
    "    serotype_data = []\n",
    "    for genome_file in genome_files:\n",
    "        records = SeqIO.parse(genome_file, \"fasta\")\n",
    "        entry = None\n",
    "        for record in records:\n",
    "            name = record.description\n",
    "            genome_name = name.split('|')[0]\n",
    "            serotype_tag = name.split('|')[1]\n",
    "            serotypes = {'O': '','H': ''}\n",
    "            for key in ['O','H']:\n",
    "                regex = re.compile(\"(?<!(Non-))(\"+key+\"\\d{1,3})(?!\\d)\")\n",
    "                results = regex.findall(serotype_tag)\n",
    "                results_len = len(results)\n",
    "                if results_len > 0:\n",
    "                    serotypes[key] = results[0][1][1:]\n",
    "            entry = {\n",
    "                'genome name': genome_name,\n",
    "                'given O': serotypes['O'],\n",
    "                'given H': serotypes['H'],\n",
    "                'serotype tag': serotype_tag\n",
    "            }\n",
    "            if serotypes['O'] == '' and serotypes['H'] =='':\n",
    "                pass\n",
    "                # print(name)\n",
    "            break\n",
    "        if entry is not None:\n",
    "            serotype_data.append(entry)\n",
    "    df2 = pd.DataFrame(serotype_data)\n",
    "    df2.index = df2['genome name']\n",
    "    df2 = df2[['given O', 'given H']]\n",
    "\n",
    "    df2.to_csv('genome-serotype_mapping.csv')\n",
    "    '''\n",
    "    geneome_df = pd.read_csv('genome-serotype_mapping.csv', index_col='genome name', dtype={'given O':str, 'given H':str})\n",
    "    \n",
    "\n",
    "    # Merge two data frames\n",
    "    prediction_df = prediction_df.merge(geneome_df, left_index=True, right_index=True, how='inner')\n",
    "    prediction_df = prediction_df[['predicted O', 'given O', 'predicted H', 'given H', 'O source', 'H source']]\n",
    "    prediction_df.head()\n",
    "    prediction_df.fillna(value=np.nan, inplace=True)\n",
    "    return prediction_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_result(result_df):\n",
    "    df1 = result_df\n",
    "    # find conflict\n",
    "    '''\n",
    "    incorrect prediction if:\n",
    "        given is different from predicton\n",
    "        given is not None\n",
    "        predicted is not None\n",
    "    '''\n",
    "    s1 = (df1['given O'].notnull() & df1['predicted O'].notnull()) & \\\n",
    "         (df1['given O']!=df1['predicted O'])\n",
    "    s2 = (df1['given H'].notnull() & df1['predicted H'].notnull()) & \\\n",
    "         (df1['given H']!=df1['predicted H'])\n",
    "    s3 = s1 | s2\n",
    "    incorrect_df = df1[s3]\n",
    "    print('%d incorrect predictions' %incorrect_df.shape[0])\n",
    "    print(incorrect_df.head())\n",
    "    '''\n",
    "    no prediction if:\n",
    "        both predicted antigen are None\n",
    "    '''\n",
    "    s1 = df1['predicted O'].isnull() & df1['predicted H'].isnull()\n",
    "    no_df = df1[s1]\n",
    "    print('%d no predictions' %no_df.shape[0])\n",
    "    print(no_df.head())\n",
    "    '''\n",
    "    correct prediction if:\n",
    "        for each given serotype, the correct prediction is made\n",
    "    '''\n",
    "    s1=df1['given O'].isnull() | (df1['given O']==df1['predicted O'])\n",
    "    s2=df1['given H'].isnull() | (df1['given H']==df1['predicted H'])\n",
    "    s3=s1&s2\n",
    "    correct_df = df1[s3]\n",
    "    # conflict rows\n",
    "    # match rows\n",
    "    print('%d correct predictions' %correct_df.shape[0])\n",
    "    print(correct_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "774 incorrect predictions\n",
      "                predicted O given O predicted H given H  O source  H source\n",
      "genome name                                                                \n",
      "ESC_AA9715AA_AS         121     157          10     NaN       NaN       NaN\n",
      "ESC_AA8012AA_AS          26     157          11     NaN       NaN       NaN\n",
      "ESC_AA7929AA_AS          26     157          11     NaN       NaN       NaN\n",
      "ESC_AA9661AA_AS         121     157          15     NaN       NaN       NaN\n",
      "ESC_AA8510AA_AS         121     157          19     NaN       NaN       NaN\n",
      "28 no predictions\n",
      "                predicted O given O predicted H given H  O source  H source\n",
      "genome name                                                                \n",
      "ESC_BA9928AA_AS         NaN     145         NaN     NaN       NaN       NaN\n",
      "ESC_CA1039AA_AS         NaN     145         NaN     NaN       NaN       NaN\n",
      "ESC_CA0497AA_AS         NaN     145         NaN     NaN       NaN       NaN\n",
      "ESC_FA8178AA_AS         NaN     112         NaN     NaN       NaN       NaN\n",
      "ESC_GA2553AA_AS         NaN     145         NaN     NaN       NaN       NaN\n",
      "4275 correct predictions\n",
      "                predicted O given O predicted H given H  O source  H source\n",
      "genome name                                                                \n",
      "ESC_AA9722AA_AS          26      26          11     NaN       NaN       NaN\n",
      "ESC_AA9419AA_AS          26      26          11      11       NaN       NaN\n",
      "ESC_AA9366AA_AS         121     121          19     NaN       NaN       NaN\n",
      "ESC_AA9052AA_AS         111     111           2     NaN       NaN       NaN\n",
      "ESC_AA7966AA_AS         113     113          21      21       NaN       NaN\n"
     ]
    }
   ],
   "source": [
    "# old allele library\n",
    "summarize_result(clean_prediction_df(pd.read_json('legacy.json')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "762 incorrect predictions\n",
      "                predicted O given O predicted H given H  O source  H source\n",
      "genome name                                                                \n",
      "ESC_AA9715AA_AS         121     157          10     NaN       NaN       NaN\n",
      "ESC_AA8012AA_AS          26     157          11     NaN       NaN       NaN\n",
      "ESC_AA7929AA_AS          26     157          11     NaN       NaN       NaN\n",
      "ESC_AA9661AA_AS         121     157          15     NaN       NaN       NaN\n",
      "ESC_AA8510AA_AS         121     157          19     NaN       NaN       NaN\n",
      "31 no predictions\n",
      "                predicted O given O predicted H given H  O source  H source\n",
      "genome name                                                                \n",
      "ESC_BA9928AA_AS         NaN     145         NaN     NaN       NaN       NaN\n",
      "ESC_CA1039AA_AS         NaN     145         NaN     NaN       NaN       NaN\n",
      "ESC_CA0497AA_AS         NaN     145         NaN     NaN       NaN       NaN\n",
      "ESC_FA8178AA_AS         NaN     112         NaN     NaN       NaN       NaN\n",
      "ESC_GA1520AA_AS         NaN     145         NaN     NaN       NaN       NaN\n",
      "4235 correct predictions\n",
      "                predicted O given O predicted H given H  O source  H source\n",
      "genome name                                                                \n",
      "ESC_AA9722AA_AS          26      26          11     NaN       NaN       NaN\n",
      "ESC_AA9419AA_AS          26      26          11      11       NaN       NaN\n",
      "ESC_AA9366AA_AS         121     121          19     NaN       NaN       NaN\n",
      "ESC_AA9052AA_AS         111     111           2     NaN       NaN       NaN\n",
      "ESC_AA7966AA_AS         113     113          21      21       NaN       NaN\n"
     ]
    }
   ],
   "source": [
    "# old allele library + 97% length cutoff\n",
    "summarize_result(clean_prediction_df(pd.read_json('legacy2.json')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "781 incorrect predictions\n",
      "                predicted O given O predicted H given H  \\\n",
      "genome name                                               \n",
      "ESC_AA9715AA_AS         121     157          10     NaN   \n",
      "ESC_AA8012AA_AS          26     157          11     NaN   \n",
      "ESC_AA7929AA_AS          26     157          11     NaN   \n",
      "ESC_AA9661AA_AS         121     157          15     NaN   \n",
      "ESC_AA8510AA_AS         121     157          19     NaN   \n",
      "\n",
      "                                                          O source  \\\n",
      "genome name                                                          \n",
      "ESC_AA9715AA_AS  8__wzx__wzx-O121__141 AY208937.1;O antigen fli...   \n",
      "ESC_AA8012AA_AS  8__wzx__wzx-O26__234 AF529080.1;O antigen flip...   \n",
      "ESC_AA7929AA_AS  8__wzx__wzx-O26__234 AF529080.1;O antigen flip...   \n",
      "ESC_AA9661AA_AS  8__wzx__wzx-O121__141 AY208937.1;O antigen fli...   \n",
      "ESC_AA8510AA_AS  8__wzx__wzx-O121__141 AY208937.1;O antigen fli...   \n",
      "\n",
      "                                H source  \n",
      "genome name                               \n",
      "ESC_AA9715AA_AS  part_of_ESC_HA8316AA_AS  \n",
      "ESC_AA8012AA_AS  part_of_ESC_KA5982AA_AS  \n",
      "ESC_AA7929AA_AS  part_of_ESC_KA5982AA_AS  \n",
      "ESC_AA9661AA_AS  part_of_ESC_HA8354AA_AS  \n",
      "ESC_AA8510AA_AS  part_of_ESC_IA2145AA_AS  \n",
      "0 no predictions\n",
      "Empty DataFrame\n",
      "Columns: [predicted O, given O, predicted H, given H, O source, H source]\n",
      "Index: []\n",
      "4337 correct predictions\n",
      "                predicted O given O predicted H given H  \\\n",
      "genome name                                               \n",
      "ESC_AA9722AA_AS          26      26          11     NaN   \n",
      "ESC_AA9419AA_AS          26      26          11      11   \n",
      "ESC_AA9366AA_AS         121     121          19     NaN   \n",
      "ESC_AA9052AA_AS         111     111           2     NaN   \n",
      "ESC_AA7966AA_AS         113     113          21      21   \n",
      "\n",
      "                                                          O source  \\\n",
      "genome name                                                          \n",
      "ESC_AA9722AA_AS  8__wzx__wzx-O26__234 AF529080.1;O antigen flip...   \n",
      "ESC_AA9419AA_AS  8__wzx__wzx-O26__234 AF529080.1;O antigen flip...   \n",
      "ESC_AA9366AA_AS  8__wzx__wzx-O121__141 AY208937.1;O antigen fli...   \n",
      "ESC_AA9052AA_AS  8__wzx__wzx-O111__128 AF078736.1;O antigen fli...   \n",
      "ESC_AA7966AA_AS                            part_of_ESC_GA3644AA_AS   \n",
      "\n",
      "                                                   H source  \n",
      "genome name                                                  \n",
      "ESC_AA9722AA_AS                     part_of_ESC_KA5982AA_AS  \n",
      "ESC_AA9419AA_AS                     part_of_ESC_KA5982AA_AS  \n",
      "ESC_AA9366AA_AS                     part_of_ESC_IA2145AA_AS  \n",
      "ESC_AA9052AA_AS  1__fliC__fliC-H2__21 HQ116826;flagellin;H2  \n",
      "ESC_AA7966AA_AS                     part_of_ESC_HA9063AA_AS  \n"
     ]
    }
   ],
   "source": [
    "# new allele library\n",
    "summarize_result(clean_prediction_df(pd.read_json('new.json')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "767 incorrect predictions\n",
      "                predicted O given O predicted H given H  \\\n",
      "genome name                                               \n",
      "ESC_AA9715AA_AS         121     157          10     NaN   \n",
      "ESC_AA8012AA_AS          26     157          11     NaN   \n",
      "ESC_AA7929AA_AS          26     157          11     NaN   \n",
      "ESC_AA9661AA_AS         121     157          15     NaN   \n",
      "ESC_AA8510AA_AS         121     157          19     NaN   \n",
      "\n",
      "                                                          O source  \\\n",
      "genome name                                                          \n",
      "ESC_AA9715AA_AS  8__wzx__wzx-O121__141 AY208937.1;O antigen fli...   \n",
      "ESC_AA8012AA_AS  8__wzx__wzx-O26__234 AF529080.1;O antigen flip...   \n",
      "ESC_AA7929AA_AS  8__wzx__wzx-O26__234 AF529080.1;O antigen flip...   \n",
      "ESC_AA9661AA_AS  8__wzx__wzx-O121__141 AY208937.1;O antigen fli...   \n",
      "ESC_AA8510AA_AS  8__wzx__wzx-O121__141 AY208937.1;O antigen fli...   \n",
      "\n",
      "                                H source  \n",
      "genome name                               \n",
      "ESC_AA9715AA_AS  part_of_ESC_HA8316AA_AS  \n",
      "ESC_AA8012AA_AS  part_of_ESC_KA5982AA_AS  \n",
      "ESC_AA7929AA_AS  part_of_ESC_KA5982AA_AS  \n",
      "ESC_AA9661AA_AS  part_of_ESC_HA8354AA_AS  \n",
      "ESC_AA8510AA_AS  part_of_ESC_IA2145AA_AS  \n",
      "0 no predictions\n",
      "Empty DataFrame\n",
      "Columns: [predicted O, given O, predicted H, given H, O source, H source]\n",
      "Index: []\n",
      "4293 correct predictions\n",
      "                predicted O given O predicted H given H  \\\n",
      "genome name                                               \n",
      "ESC_AA9722AA_AS          26      26          11     NaN   \n",
      "ESC_AA9419AA_AS          26      26          11      11   \n",
      "ESC_AA9366AA_AS         121     121          19     NaN   \n",
      "ESC_AA9052AA_AS         111     111           2     NaN   \n",
      "ESC_AA7966AA_AS         113     113          21      21   \n",
      "\n",
      "                                                          O source  \\\n",
      "genome name                                                          \n",
      "ESC_AA9722AA_AS  8__wzx__wzx-O26__234 AF529080.1;O antigen flip...   \n",
      "ESC_AA9419AA_AS  8__wzx__wzx-O26__234 AF529080.1;O antigen flip...   \n",
      "ESC_AA9366AA_AS  8__wzx__wzx-O121__141 AY208937.1;O antigen fli...   \n",
      "ESC_AA9052AA_AS  8__wzx__wzx-O111__128 AF078736.1;O antigen fli...   \n",
      "ESC_AA7966AA_AS                            part_of_ESC_GA3644AA_AS   \n",
      "\n",
      "                                                   H source  \n",
      "genome name                                                  \n",
      "ESC_AA9722AA_AS                     part_of_ESC_KA5982AA_AS  \n",
      "ESC_AA9419AA_AS                     part_of_ESC_KA5982AA_AS  \n",
      "ESC_AA9366AA_AS                     part_of_ESC_IA2145AA_AS  \n",
      "ESC_AA9052AA_AS  1__fliC__fliC-H2__21 HQ116826;flagellin;H2  \n",
      "ESC_AA7966AA_AS                     part_of_ESC_HA9063AA_AS  \n"
     ]
    }
   ],
   "source": [
    "# new allele library + 97% length cutoff\n",
    "summarize_result(clean_prediction_df(pd.read_json('new2.json')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
