{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def clean_prediction_df(prediction_df):\n",
    "    '''\n",
    "    # Generate Allele Mapping\n",
    "    serotype_dict = json.load(open('serotype_dict.json'))\n",
    "    new_list = []\n",
    "    for serotype, alleles in serotype_dict.items():\n",
    "        for allele in alleles:\n",
    "            new_entry = {\n",
    "                'H source': allele.get('name'),\n",
    "                'H allele': '-'.join([serotype, str(allele.get('num'))])\n",
    "            }\n",
    "            new_list.append(new_entry)\n",
    "    allele_df = pd.DataFrame(new_list)\n",
    "    allele_df.to_csv('allele_mapping.csv', index_label=False)\n",
    "    '''\n",
    "    allele_df = pd.read_csv('allele_mapping.csv')\n",
    "\n",
    "    # Merge with allele data\n",
    "    prediction_df = prediction_df.merge(allele_df, left_on='H allele', right_on='H allele', how='left')\n",
    "    allele_df.columns = ['O allele', 'O source']\n",
    "    prediction_df = prediction_df.merge(allele_df, left_on='O allele', right_on='O allele', how='left')\n",
    "    prediction_df.head()\n",
    "\n",
    "    # Read ectyper output\n",
    "    # Data cleaning\n",
    "    prediction_df['genome name'] = prediction_df['genome name'].str.split('|', expand=True)[1]\n",
    "    prediction_df['predicted O'] = prediction_df['predicted O'].str.split('O', expand=True)[1]\n",
    "    prediction_df['predicted H'] = prediction_df['predicted H'].str.split('H', expand=True)[1]\n",
    "    \n",
    "    '''\n",
    "    # Get genome file paths as a list\n",
    "    import os\n",
    "    genome_dir = '/home/sam/Projects/MoreSerotype/temp/genomes'\n",
    "    genome_names = os.listdir(genome_dir)\n",
    "    genome_files = []\n",
    "    for genome_name in genome_names:\n",
    "        file = os.path.join(genome_dir, genome_name)\n",
    "        genome_files.append(file)\n",
    "\n",
    "    # Create serotype dataframe\n",
    "    from Bio import SeqIO\n",
    "    import re\n",
    "    serotype_data = []\n",
    "    for genome_file in genome_files:\n",
    "        records = SeqIO.parse(genome_file, \"fasta\")\n",
    "        entry = None\n",
    "        for record in records:\n",
    "            name = record.description\n",
    "            genome_name = name.split('|')[0]\n",
    "            serotype_tag = name.split('|')[1]\n",
    "            serotypes = {'O': '','H': ''}\n",
    "            for key in ['O','H']:\n",
    "                regex = re.compile(\"(?<!(Non-))(\"+key+\"\\d{1,3})(?!\\d)\")\n",
    "                results = regex.findall(serotype_tag)\n",
    "                results_len = len(results)\n",
    "                if results_len > 0:\n",
    "                    serotypes[key] = results[0][1][1:]\n",
    "            entry = {\n",
    "                'genome name': genome_name,\n",
    "                'given O': serotypes['O'],\n",
    "                'given H': serotypes['H'],\n",
    "                'serotype tag': serotype_tag\n",
    "            }\n",
    "            if serotypes['O'] == '' and serotypes['H'] =='':\n",
    "                pass\n",
    "                # print(name)\n",
    "            break\n",
    "        if entry is not None:\n",
    "            serotype_data.append(entry)\n",
    "    df2 = pd.DataFrame(serotype_data)\n",
    "    df2.index = df2['genome name']\n",
    "    df2 = df2[['given O', 'given H']]\n",
    "\n",
    "    df2.to_csv('genome-serotype_mapping.csv')\n",
    "    '''\n",
    "    geneome_df = pd.read_csv('genome-serotype_mapping.csv', index_col='genome name', dtype={'given O':str, 'given H':str})\n",
    "    \n",
    "\n",
    "    # Merge two data frames\n",
    "    prediction_df = prediction_df.merge(geneome_df, left_on='genome name', right_index=True, how='inner')\n",
    "    \n",
    "    prediction_df = prediction_df.replace('-', np.nan)\n",
    "    prediction_df.fillna(value=np.nan, inplace=True)\n",
    "    \n",
    "    # Rename columns\n",
    "    prediction_df.columns = [\n",
    "        'H_allele', 'O_allele', 'genome_name',\n",
    "        'predicted_H', 'predicted_O', 'H_source',\n",
    "        'O_source', 'given_O', 'given_H'\n",
    "    ]\n",
    "    # Reorder columns\n",
    "    prediction_df = prediction_df[[\n",
    "        'genome_name',\n",
    "        'given_O', 'predicted_O',\n",
    "        'given_H', 'predicted_H',\n",
    "        'O_allele', 'O_source',\n",
    "        'H_allele', 'H_source'\n",
    "    ]]\n",
    "    return prediction_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def summarize_result(result_df):\n",
    "    important_cols = result_df.columns\n",
    "    \n",
    "    df = result_df\n",
    "    print('all predictions(%d):' %result_df.shape[0])\n",
    "    '''\n",
    "    no prediction if:\n",
    "        both predicted antigen are None\n",
    "    '''\n",
    "    s1 = (df['predicted_O'].isnull() & df['predicted_H'].isnull())\n",
    "    no_df = df[s1][important_cols]\n",
    "    print('%d no predictions' %no_df.shape[0])\n",
    "    '''\n",
    "    incorrect prediction if:\n",
    "        not in 'no_df'\n",
    "        at least one prediction is wrong\n",
    "    '''\n",
    "    s2 = (\n",
    "        ((df['given_O'].notnull() & df['predicted_O'].notnull()) & (df['given_O']!=df['predicted_O'])) |\n",
    "        ((df['given_H'].notnull() & df['predicted_H'].notnull()) & (df['given_H']!=df['predicted_H']))\n",
    "    )\n",
    "    incorrect_df = df[~s1 & s2][important_cols]\n",
    "    print('%d incorrect predictions' %incorrect_df.shape[0])\n",
    "    '''\n",
    "    correct prediction if:\n",
    "        not in 'no_df'\n",
    "        not in 'incorrect_df'\n",
    "        for each given serotype, the correct prediction is made\n",
    "    '''\n",
    "    s3 = (\n",
    "        (df['given_O'].isnull() | (df['given_O']==df['predicted_O'])) &\n",
    "        (df['given_H'].isnull() | (df['given_H']==df['predicted_H']))\n",
    "    )\n",
    "    correct_df = df[~s1 & ~s2 & s3][important_cols]\n",
    "    print('%d correct predictions' %correct_df.shape[0])\n",
    "    '''\n",
    "    semicorrect prediction if:\n",
    "        not in 'no_df'\n",
    "        not in 'incorrect_df'\n",
    "        not in 'correct_df'\n",
    "        one correct prediction is made, no prediction for the other\n",
    "    '''\n",
    "    s4 = (\n",
    "        (df['given_O'].isnull() | (df['given_O']==df['predicted_O'])) |\n",
    "        (df['given_H'].isnull() | (df['given_H']==df['predicted_H']))\n",
    "    )\n",
    "    semicorrect_df = df[~s1 & ~s2 & ~s3 & s4][important_cols]\n",
    "    print('%d semicorrect predictions' %semicorrect_df.shape[0])\n",
    "    '''\n",
    "    remaining prediction\n",
    "    '''\n",
    "    remaining_df = df[~s1 & ~s2 & ~s3 & ~s4]\n",
    "    print('%d remaining predictions' %remaining_df.shape[0])\n",
    "    \n",
    "    # Summary\n",
    "    \n",
    "    # O serotype\n",
    "    # get number of given serotypes\n",
    "    total_s = df['given_O'].notnull()\n",
    "    num_total = df[total_s].shape[0]\n",
    "    print(\"number of given_O serotypes is %d\"\n",
    "            %num_total)\n",
    "    # get number of unpredicted serotype\n",
    "    unpredicted_s = total_s & df['predicted_O'].isnull()\n",
    "    num_unpredicted = (df[unpredicted_s].shape[0])\n",
    "    print(\"number of unpredicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_unpredicted, num_unpredicted/num_total*100))\n",
    "    correct_s = total_s & ~unpredicted_s & (df['given_O']==df['predicted_O'])\n",
    "    num_correct = df[correct_s].shape[0]\n",
    "    print(\"number of correctly predicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_correct, num_correct/num_total*100))\n",
    "    incorrect_s = total_s & ~unpredicted_s & ~correct_s\n",
    "    incorrect_o_df = df[incorrect_s][important_cols]\n",
    "    num_incorrect = incorrect_o_df.shape[0]\n",
    "    print(\"number of incorrectly predicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_incorrect, num_incorrect/num_total*100))\n",
    "    \n",
    "    # H serotype\n",
    "    # get number of given serotypes\n",
    "    total_s = df['given_H'].notnull()\n",
    "    num_total = (df[total_s].shape[0])\n",
    "    print(\"number of given_H serotypes is %d\"\n",
    "            %num_total)\n",
    "    # get number of unpredicted serotype\n",
    "    unpredicted_s = total_s & df['predicted_H'].isnull()\n",
    "    num_unpredicted = (df[unpredicted_s].shape[0])\n",
    "    print(\"number of unpredicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_unpredicted, num_unpredicted/num_total*100))\n",
    "    correct_s = total_s & ~unpredicted_s & (df['given_H']==df['predicted_H'])\n",
    "    num_correct = df[correct_s].shape[0]\n",
    "    print(\"number of correctly predicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_correct, num_correct/num_total*100))\n",
    "    incorrect_s = total_s & ~unpredicted_s & ~correct_s\n",
    "    incorrect_h_df = df[incorrect_s][important_cols]\n",
    "    num_incorrect = incorrect_h_df.shape[0]\n",
    "    print(\"number of incorrectly predicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_incorrect, num_incorrect/num_total*100))\n",
    "\n",
    "    return no_df, incorrect_df, correct_df, semicorrect_df, incorrect_o_df, incorrect_h_df, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all predictions(5810):\n",
      "28 no predictions\n",
      "774 incorrect predictions\n",
      "4275 correct predictions\n",
      "733 semicorrect predictions\n",
      "0 remaining predictions\n",
      "number of given_O serotypes is 5671\n",
      "number of unpredicted_O serotypes is 867 or 15.29%\n",
      "number of correctly predicted_O serotypes is 4326 or 76.28%\n",
      "number of incorrectly predicted_O serotypes is 478 or 8.43%\n",
      "number of given_H serotypes is 2887\n",
      "number of unpredicted_H serotypes is 50 or 1.73%\n",
      "number of correctly predicted_H serotypes is 2433 or 84.27%\n",
      "number of incorrectly predicted_H serotypes is 404 or 13.99%\n"
     ]
    }
   ],
   "source": [
    "# old allele library\n",
    "raw_df = pd.read_json('legacy.json')\n",
    "clean_df = clean_prediction_df(raw_df)\n",
    "result1 = summarize_result(clean_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all predictions(5810):\n",
      "31 no predictions\n",
      "762 incorrect predictions\n",
      "4235 correct predictions\n",
      "782 semicorrect predictions\n",
      "0 remaining predictions\n",
      "number of given_O serotypes is 5671\n",
      "number of unpredicted_O serotypes is 925 or 16.31%\n",
      "number of correctly predicted_O serotypes is 4282 or 75.51%\n",
      "number of incorrectly predicted_O serotypes is 464 or 8.18%\n",
      "number of given_H serotypes is 2887\n",
      "number of unpredicted_H serotypes is 52 or 1.80%\n",
      "number of correctly predicted_H serotypes is 2432 or 84.24%\n",
      "number of incorrectly predicted_H serotypes is 403 or 13.96%\n"
     ]
    }
   ],
   "source": [
    "# old allele library + 97% length cutoff\n",
    "raw_df = pd.read_json('legacy2.json')\n",
    "clean_df = clean_prediction_df(raw_df)\n",
    "result2 = summarize_result(clean_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all predictions(5792):\n",
      "0 no predictions\n",
      "781 incorrect predictions\n",
      "4337 correct predictions\n",
      "674 semicorrect predictions\n",
      "0 remaining predictions\n",
      "number of given_O serotypes is 5653\n",
      "number of unpredicted_O serotypes is 780 or 13.80%\n",
      "number of correctly predicted_O serotypes is 4379 or 77.46%\n",
      "number of incorrectly predicted_O serotypes is 494 or 8.74%\n",
      "number of given_H serotypes is 2879\n",
      "number of unpredicted_H serotypes is 28 or 0.97%\n",
      "number of correctly predicted_H serotypes is 2451 or 85.13%\n",
      "number of incorrectly predicted_H serotypes is 400 or 13.89%\n"
     ]
    }
   ],
   "source": [
    "# new allele library\n",
    "raw_df = pd.read_json('new.json')\n",
    "clean_df = clean_prediction_df(raw_df)\n",
    "result3 = summarize_result(clean_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all predictions(5789):\n",
      "0 no predictions\n",
      "767 incorrect predictions\n",
      "4293 correct predictions\n",
      "729 semicorrect predictions\n",
      "0 remaining predictions\n",
      "number of given_O serotypes is 5650\n",
      "number of unpredicted_O serotypes is 841 or 14.88%\n",
      "number of correctly predicted_O serotypes is 4331 or 76.65%\n",
      "number of incorrectly predicted_O serotypes is 478 or 8.46%\n",
      "number of given_H serotypes is 2879\n",
      "number of unpredicted_H serotypes is 30 or 1.04%\n",
      "number of correctly predicted_H serotypes is 2450 or 85.10%\n",
      "number of incorrectly predicted_H serotypes is 399 or 13.86%\n"
     ]
    }
   ],
   "source": [
    "# new allele library + 97% length cutoff\n",
    "raw_df = pd.read_json('new2.json')\n",
    "clean_df = clean_prediction_df(raw_df)\n",
    "result4 = summarize_result(clean_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all predictions(5789):\n",
      "17 no predictions\n",
      "674 incorrect predictions\n",
      "3706 correct predictions\n",
      "1392 semicorrect predictions\n",
      "0 remaining predictions\n",
      "number of given_O serotypes is 5650\n",
      "number of unpredicted_O serotypes is 1577 or 27.91%\n",
      "number of correctly predicted_O serotypes is 3710 or 65.66%\n",
      "number of incorrectly predicted_O serotypes is 363 or 6.42%\n",
      "number of given_H serotypes is 2879\n",
      "number of unpredicted_H serotypes is 30 or 1.04%\n",
      "number of correctly predicted_H serotypes is 2450 or 85.10%\n",
      "number of incorrectly predicted_H serotypes is 399 or 13.86%\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_name</th>\n",
       "      <th>given_O</th>\n",
       "      <th>predicted_O</th>\n",
       "      <th>given_H</th>\n",
       "      <th>predicted_H</th>\n",
       "      <th>O_allele</th>\n",
       "      <th>O_source</th>\n",
       "      <th>H_allele</th>\n",
       "      <th>H_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1657</th>\n",
       "      <td>ESC_GA6847AA_AS</td>\n",
       "      <td>151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H4-2</td>\n",
       "      <td>1__fliC__fliC-H4__43 AJ605764.1;flagellin;H4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1808</th>\n",
       "      <td>ESC_GA8882AA_AS</td>\n",
       "      <td>151</td>\n",
       "      <td>118</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>O118-2</td>\n",
       "      <td>9__wzy__wzy-O118-Gp3__341 HM204927.1;O antigen...</td>\n",
       "      <td>H11-18</td>\n",
       "      <td>part_of_ESC_GA8882AA_AS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1810</th>\n",
       "      <td>ESC_GA8836AA_AS</td>\n",
       "      <td>151</td>\n",
       "      <td>118</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>O118-2</td>\n",
       "      <td>9__wzy__wzy-O118-Gp3__341 HM204927.1;O antigen...</td>\n",
       "      <td>H11-18</td>\n",
       "      <td>part_of_ESC_GA8882AA_AS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1909</th>\n",
       "      <td>ESC_GA8952AA_AS</td>\n",
       "      <td>151</td>\n",
       "      <td>118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>O118-2</td>\n",
       "      <td>9__wzy__wzy-O118-Gp3__341 HM204927.1;O antigen...</td>\n",
       "      <td>H11-18</td>\n",
       "      <td>part_of_ESC_GA8882AA_AS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2200</th>\n",
       "      <td>ESC_GA9397AA_AS</td>\n",
       "      <td>151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H11-18</td>\n",
       "      <td>part_of_ESC_GA8882AA_AS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2387</th>\n",
       "      <td>ESC_HA1807AA_AS</td>\n",
       "      <td>151</td>\n",
       "      <td>118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>O118-2</td>\n",
       "      <td>9__wzy__wzy-O118-Gp3__341 HM204927.1;O antigen...</td>\n",
       "      <td>H10-4</td>\n",
       "      <td>part_of_ESC_GA8794AA_AS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3522</th>\n",
       "      <td>ESC_HA8317AA_AS</td>\n",
       "      <td>151</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30</td>\n",
       "      <td>O10-4</td>\n",
       "      <td>part_of_ESC_HA7958AA_AS</td>\n",
       "      <td>H30-1</td>\n",
       "      <td>1__fliC__fliC-H30__31 AY250011.1;flagellin;H30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3569</th>\n",
       "      <td>ESC_HA8276AA_AS</td>\n",
       "      <td>151</td>\n",
       "      <td>78</td>\n",
       "      <td>41</td>\n",
       "      <td>9</td>\n",
       "      <td>O78-4</td>\n",
       "      <td>part_of_ESC_LA5784AA_AS</td>\n",
       "      <td>H9-1</td>\n",
       "      <td>1__fliC__fliC-H9__76 AY249994.1;flagellin;H9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          genome_name given_O predicted_O given_H predicted_H O_allele  \\\n",
       "1657  ESC_GA6847AA_AS     151         NaN       4           4      NaN   \n",
       "1808  ESC_GA8882AA_AS     151         118      11          11   O118-2   \n",
       "1810  ESC_GA8836AA_AS     151         118      11          11   O118-2   \n",
       "1909  ESC_GA8952AA_AS     151         118     NaN          11   O118-2   \n",
       "2200  ESC_GA9397AA_AS     151         NaN     NaN          11      NaN   \n",
       "2387  ESC_HA1807AA_AS     151         118     NaN          10   O118-2   \n",
       "3522  ESC_HA8317AA_AS     151          10     NaN          30    O10-4   \n",
       "3569  ESC_HA8276AA_AS     151          78      41           9    O78-4   \n",
       "\n",
       "                                               O_source H_allele  \\\n",
       "1657                                                NaN     H4-2   \n",
       "1808  9__wzy__wzy-O118-Gp3__341 HM204927.1;O antigen...   H11-18   \n",
       "1810  9__wzy__wzy-O118-Gp3__341 HM204927.1;O antigen...   H11-18   \n",
       "1909  9__wzy__wzy-O118-Gp3__341 HM204927.1;O antigen...   H11-18   \n",
       "2200                                                NaN   H11-18   \n",
       "2387  9__wzy__wzy-O118-Gp3__341 HM204927.1;O antigen...    H10-4   \n",
       "3522                            part_of_ESC_HA7958AA_AS    H30-1   \n",
       "3569                            part_of_ESC_LA5784AA_AS     H9-1   \n",
       "\n",
       "                                            H_source  \n",
       "1657    1__fliC__fliC-H4__43 AJ605764.1;flagellin;H4  \n",
       "1808                         part_of_ESC_GA8882AA_AS  \n",
       "1810                         part_of_ESC_GA8882AA_AS  \n",
       "1909                         part_of_ESC_GA8882AA_AS  \n",
       "2200                         part_of_ESC_GA8882AA_AS  \n",
       "2387                         part_of_ESC_GA8794AA_AS  \n",
       "3522  1__fliC__fliC-H30__31 AY250011.1;flagellin;H30  \n",
       "3569    1__fliC__fliC-H9__76 AY249994.1;flagellin;H9  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# oct30\n",
    "raw_df = pd.read_json('oct30_output/output.json')\n",
    "clean_df = clean_prediction_df(raw_df)\n",
    "result5 = summarize_result(clean_df)\n",
    "result5[6][result5[6].given_O=='151']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
