{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_diff_row(df1, df2, on):\n",
    "    df = df1.merge(df2, on=on, how='outer', indicator=True)\n",
    "    return (df[df._merge!='both'])\n",
    "def parse_assembly_summary(file):\n",
    "    df = pd.read_csv(file, sep='\\t', index_col='assembly_accession')\n",
    "    \n",
    "    # Get genome name\n",
    "    df['genome_name'] = pd.DataFrame(df['ftp_path'].str.split('/', expand=True).iloc[:,-1])\n",
    "\n",
    "    # Add O serotype info\n",
    "    serotype_re = '(?<![A-z])(O\\d{1,3})'\n",
    "    df['given_O'] = df['organism_name'].str.extract(serotype_re, expand=False)\n",
    "    df.given_O.fillna(df['infraspecific_name'].str.extract(serotype_re, expand=False), inplace=True)\n",
    "\n",
    "    # Add H serotype info\n",
    "    serotype_re2 = '\\:(H\\d{1,2})'\n",
    "    df['given_H'] = df['organism_name'].str.extract(serotype_re2, expand=False)\n",
    "    df.given_H.fillna(df['infraspecific_name'].str.extract(serotype_re2, expand=False), inplace=True)\n",
    "\n",
    "    # Remove rows without serotype\n",
    "    df = df[df['given_O'].notnull() | df['given_H'].notnull()].copy()\n",
    "    \n",
    "    # Filter out incorrect H serotypes\n",
    "    filter_s = (df['given_H'].str[1:].astype(float)<=56) | df['given_H'].isnull()\n",
    "    filter_s2 = (df['given_O'].str[1:].astype(float)<=186) | df['given_O'].isnull()\n",
    "    df = df[filter_s & filter_s2]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def summarize_result(result_df):\n",
    "    important_cols = [\n",
    "        'organism_name', 'genome_name', 'infraspecific_name',\n",
    "        'given_O', 'predicted_O', 'O_info',\n",
    "        'given_H', 'predicted_H', 'H_info'\n",
    "    ]\n",
    "    \n",
    "    df = result_df.copy()\n",
    "    print('all predictions(%d):' %result_df.shape[0])\n",
    "    '''\n",
    "    no prediction if:\n",
    "        both predicted antigen are None\n",
    "    '''\n",
    "    s1 = (df['predicted_O'].isnull() & df['predicted_H'].isnull())\n",
    "    no_df = df[s1][important_cols]\n",
    "    print('%d no predictions' %no_df.shape[0])\n",
    "    '''\n",
    "    incorrect prediction if:\n",
    "        not in 'no_df'\n",
    "        at least one prediction is wrong\n",
    "    '''\n",
    "    s2 = (\n",
    "        ((df['given_O'].notnull() & df['predicted_O'].notnull()) & (df['given_O']!=df['predicted_O'])) |\n",
    "        ((df['given_H'].notnull() & df['predicted_H'].notnull()) & (df['given_H']!=df['predicted_H']))\n",
    "    )\n",
    "    incorrect_df = df[~s1 & s2][important_cols]\n",
    "    print('%d incorrect predictions' %incorrect_df.shape[0])\n",
    "    '''\n",
    "    correct prediction if:\n",
    "        not in 'no_df'\n",
    "        not in 'incorrect_df'\n",
    "        for each given serotype, the correct prediction is made\n",
    "    '''\n",
    "    s3 = (\n",
    "        (df['given_O'].isnull() | (df['given_O']==df['predicted_O'])) &\n",
    "        (df['given_H'].isnull() | (df['given_H']==df['predicted_H']))\n",
    "    )\n",
    "    correct_df = df[~s1 & ~s2 & s3][important_cols]\n",
    "    print('%d correct predictions' %correct_df.shape[0])\n",
    "    '''\n",
    "    semicorrect prediction if:\n",
    "        not in 'no_df'\n",
    "        not in 'incorrect_df'\n",
    "        not in 'correct_df'\n",
    "        one correct prediction is made, no prediction for the other\n",
    "    '''\n",
    "    s4 = (\n",
    "        (df['given_O'].isnull() | (df['given_O']==df['predicted_O'])) |\n",
    "        (df['given_H'].isnull() | (df['given_H']==df['predicted_H']))\n",
    "    )\n",
    "    semicorrect_df = df[~s1 & ~s2 & ~s3 & s4][important_cols]\n",
    "    print('%d semicorrect predictions' %semicorrect_df.shape[0])\n",
    "    '''\n",
    "    remaining prediction\n",
    "    '''\n",
    "    remaining_df = df[~s1 & ~s2 & ~s3 & ~s4]\n",
    "    print('%d remaining predictions' %remaining_df.shape[0])\n",
    "    \n",
    "    # Summary\n",
    "    correct_count = 0\n",
    "    incorrect_count = 0\n",
    "    given_count = 0\n",
    "    # O serotype\n",
    "    # get number of given serotypes\n",
    "    total_s = df['given_O'].notnull()\n",
    "    num_total = df[total_s].shape[0]\n",
    "    given_count += num_total\n",
    "    print(\"number of given_O serotypes is %d\"\n",
    "            %num_total)\n",
    "    # get number of unpredicted serotype\n",
    "    unpredicted_s = total_s & df['predicted_O'].isnull()\n",
    "    num_unpredicted = (df[unpredicted_s].shape[0])\n",
    "    print(\"number of unpredicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_unpredicted, num_unpredicted/num_total*100))\n",
    "    correct_s = total_s & ~unpredicted_s & (df['given_O']==df['predicted_O'])\n",
    "    num_correct = df[correct_s].shape[0]\n",
    "    correct_count += num_correct\n",
    "    print(\"number of correctly predicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_correct, num_correct/num_total*100))\n",
    "    incorrect_s = total_s & ~unpredicted_s & ~correct_s\n",
    "    incorrect_o_df = df[incorrect_s][important_cols]\n",
    "    num_incorrect = incorrect_o_df.shape[0]\n",
    "    incorrect_count += num_incorrect\n",
    "    print(\"number of incorrectly predicted_O serotypes is %d or %.2f%%\"\n",
    "            %(num_incorrect, num_incorrect/num_total*100))\n",
    "    \n",
    "    # H serotype\n",
    "    # get number of given serotypes\n",
    "    total_s = df['given_H'].notnull()\n",
    "    num_total = (df[total_s].shape[0])\n",
    "    given_count += num_total\n",
    "    print(\"number of given_H serotypes is %d\"\n",
    "            %num_total)\n",
    "    # get number of unpredicted serotype\n",
    "    unpredicted_s = total_s & df['predicted_H'].isnull()\n",
    "    num_unpredicted = (df[unpredicted_s].shape[0])\n",
    "    print(\"number of unpredicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_unpredicted, num_unpredicted/num_total*100))\n",
    "    correct_s = total_s & ~unpredicted_s & (df['given_H']==df['predicted_H'])\n",
    "    num_correct = df[correct_s].shape[0]\n",
    "    correct_count += num_correct\n",
    "    print(\"number of correctly predicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_correct, num_correct/num_total*100))\n",
    "    incorrect_s = total_s & ~unpredicted_s & ~correct_s\n",
    "    incorrect_h_df = df[incorrect_s][important_cols]\n",
    "    num_incorrect = incorrect_h_df.shape[0]\n",
    "    incorrect_count += num_incorrect\n",
    "    print(\"number of incorrectly predicted_H serotypes is %d or %.2f%%\"\n",
    "            %(num_incorrect, num_incorrect/num_total*100))\n",
    "    print(\"Overall concordance=%.2f%%(%d/%d)\" %(correct_count/given_count*100, correct_count,given_count))\n",
    "    print(\"Overall discrepancies=%.2f%%(%d/%d)\" %(incorrect_count/given_count*100, incorrect_count, given_count))\n",
    "\n",
    "    return no_df, incorrect_df, correct_df, semicorrect_df, incorrect_o_df, incorrect_h_df, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def result_to_summary(df):\n",
    "    h_interested_cols = ['genome_name', 'given_H', 'predicted_H', 'H_info', 'fliC', 'fllA', 'flkA', 'flmA', 'flnA']\n",
    "    o_interested_cols = ['genome_name', 'given_O', 'predicted_O', 'O_info', 'wzx', 'wzy', 'wzm', 'wzt']\n",
    "    # split into two dfs\n",
    "    def rename_antigen_cols(df, antigen):\n",
    "        df = df.rename(columns={\n",
    "            'given_'+antigen: 'given',\n",
    "            'predicted_'+antigen:'predicted',\n",
    "            antigen+'_info':'info'})\n",
    "        df = df[df['given'].notnull() | df['predicted'].notnull()]\n",
    "        df = df[df['given'] != df['predicted']]\n",
    "        return df\n",
    "    mislabelled_genomes = ['GCA_000695155_1', 'GCA_000234315_2', 'GCA_000671295_1', 'GCA_000520035_1', 'GCA_001940375_1', 'GCA_000285655_3', 'GCA_000617365_2', 'GCA_000617425_1', 'GCA_000619345_1', 'GCA_000632575_1', 'GCA_000617045_2']\n",
    "    h_df = rename_antigen_cols(df[h_interested_cols], 'H')\n",
    "    o_df = rename_antigen_cols(df[o_interested_cols], 'O')\n",
    "    df = h_df.append(o_df, ignore_index=True)\n",
    "    interested_cols = ['genome_name', 'given', 'predicted', 'info']\n",
    "    df = df[interested_cols]\n",
    "    df['explanation'] = np.nan\n",
    "    df.loc[(df.genome_name.apply(lambda x: '_'.join(x.replace('.', '_').split('_')[:3])).isin(mislabelled_genomes)) & df.explanation.isnull(), 'explanation'] = 'potentially mislabelled'\n",
    "    df.loc[(df['info']=='No alignment found') & df.explanation.isnull(), 'explanation'] = 'No matching alignment'\n",
    "    df.loc[df.given.isnull() & df.explanation.isnull(), 'explanation'] = 'Insufficient serotype info'\n",
    "    df.loc[(df.genome_name=='GCA_000617165.2_Ec02-3404') & df.explanation.isnull(), 'explanation'] = 'identified as non-ecoli by reference marker alignment'\n",
    "    df.loc[df['info'].isnull() & df.explanation.isnull(), 'explanation'] = 'File system error'\n",
    "    df.loc[df.explanation.isnull(), 'explanation'] = 'Not sure'\n",
    "    df = df.drop(['info'], axis=1)\n",
    "    df.to_csv('genbank_result_summary.csv', na_rep='N/A')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(536, 9)\n",
      "all predictions(536):\n",
      "1 no predictions\n",
      "8 incorrect predictions\n",
      "488 correct predictions\n",
      "39 semicorrect predictions\n",
      "0 remaining predictions\n",
      "number of given_O serotypes is 535\n",
      "number of unpredicted_O serotypes is 26 or 4.86%\n",
      "number of correctly predicted_O serotypes is 504 or 94.21%\n",
      "number of incorrectly predicted_O serotypes is 5 or 0.93%\n",
      "number of given_H serotypes is 387\n",
      "number of unpredicted_H serotypes is 15 or 3.88%\n",
      "number of correctly predicted_H serotypes is 368 or 95.09%\n",
      "number of incorrectly predicted_H serotypes is 4 or 1.03%\n",
      "Overall concordance=94.58%(872/922)\n",
      "Overall discrepancies=0.98%(9/922)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['GCA_000147855.3_ASM14785v3',\n",
       " 'GCA_000258025.1_ASM25802v1',\n",
       " 'GCA_000281775.1_ASM28177v1',\n",
       " 'GCA_001309885.1_ASM130988v1',\n",
       " 'GCA_001309895.1_ASM130989v1',\n",
       " 'GCA_001309905.1_ASM130990v1',\n",
       " 'GCA_001309925.1_ASM130992v1',\n",
       " 'GCA_001309965.1_ASM130996v1',\n",
       " 'GCA_001440355.1_ECO5_020',\n",
       " 'GCA_001440565.1_ASM144056v1',\n",
       " 'GCA_001440615.1_ASM144061v1',\n",
       " 'GCA_001440645.1_ASM144064v1',\n",
       " 'GCA_001440655.1_ASM144065v1',\n",
       " 'GCA_001440665.1_ASM144066v1',\n",
       " 'GCA_001637865.1_ASM163786v1',\n",
       " 'GCA_001638005.1_ASM163800v1',\n",
       " 'GCA_001700405.1_ASM170040v1',\n",
       " 'GCA_002109305.1_E_coli_SHECO001_v1',\n",
       " 'GCA_002111485.1_E_coli_SHECO002_v1',\n",
       " 'GCA_002111495.1_E_coli_SHECO003_v1',\n",
       " 'GCA_002215095.1_ASM221509v1',\n",
       " 'GCA_002215115.1_ASM221511v1',\n",
       " 'GCA_002215155.1_ASM221515v1',\n",
       " 'GCA_002317635.1_ASM231763v1',\n",
       " 'GCA_002317645.1_ASM231764v1',\n",
       " 'GCA_002317655.1_ASM231765v1',\n",
       " 'GCA_002317695.1_ASM231769v1',\n",
       " 'GCA_002317705.1_ASM231770v1',\n",
       " 'GCA_002317725.1_ASM231772v1',\n",
       " 'GCA_002317755.1_ASM231775v1',\n",
       " 'GCA_002317775.1_ASM231777v1']"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_df = parse_assembly_summary('assembly_summary.csv')[['genome_name', 'organism_name', 'infraspecific_name','given_O', 'given_H']]\n",
    "output_df = pd.read_csv('output/genbank/output.csv')\n",
    "output_df['genome_name']=output_df['index'].str.split('_genomic').str[0]\n",
    "output_df.drop('index', axis=1, inplace=True)\n",
    "output_df = output_df[['O_prediction', 'O_info', 'H_prediction', 'H_info', 'genome_name']]\n",
    "output_df.columns = ['predicted_O', 'O_info', 'predicted_H', 'H_info', 'genome_name']\n",
    "output_df.loc[output_df.predicted_O=='-', 'predicted_O'] = np.nan\n",
    "output_df.loc[output_df.predicted_H=='-', 'predicted_H'] = np.nan\n",
    "merge_df = mapping_df.merge(output_df, on='genome_name', how='inner')\n",
    "print(merge_df.shape)\n",
    "results = summarize_result(merge_df)\n",
    "results[1]\n",
    "# Remove the genomes below from folder\n",
    "diff_df = get_diff_row(mapping_df, output_df, 'genome_name')\n",
    "list(diff_df.genome_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>genbank provided</th>\n",
       "      <th>analysis</th>\n",
       "      <th>ectyper predicted</th>\n",
       "      <th>new predicted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>GCA_000749525_lcl.fasta</td>\n",
       "      <td>O91H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O181:H49</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>GCA_000749545_lcl.fasta</td>\n",
       "      <td>O91H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O91:H21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>GCA_000749565_lcl.fasta</td>\n",
       "      <td>O91H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O91:H21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>GCA_000749575_lcl.fasta</td>\n",
       "      <td>O91H21</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O91:H21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>GCA_000819645_lcl.fasta</td>\n",
       "      <td>O78</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O78:H19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>GCA_001268885_lcl.fasta</td>\n",
       "      <td>O74</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>O66:H45</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>GCA_001268925_lcl.fasta</td>\n",
       "      <td>O76</td>\n",
       "      <td>Serotype</td>\n",
       "      <td>-:H34</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file genbank provided  analysis ectyper predicted  \\\n",
       "406  GCA_000749525_lcl.fasta           O91H21  Serotype          O181:H49   \n",
       "407  GCA_000749545_lcl.fasta           O91H21  Serotype           O91:H21   \n",
       "408  GCA_000749565_lcl.fasta           O91H21  Serotype           O91:H21   \n",
       "409  GCA_000749575_lcl.fasta           O91H21  Serotype           O91:H21   \n",
       "411  GCA_000819645_lcl.fasta              O78  Serotype           O78:H19   \n",
       "421  GCA_001268885_lcl.fasta              O74  Serotype           O66:H45   \n",
       "422  GCA_001268925_lcl.fasta              O76  Serotype             -:H34   \n",
       "\n",
       "    new predicted  \n",
       "406           NaN  \n",
       "407           NaN  \n",
       "408           NaN  \n",
       "409           NaN  \n",
       "411           NaN  \n",
       "421           NaN  \n",
       "422           NaN  "
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_excel('in_silico_summary.xlsx')\n",
    "df = df[df.file.notnull()]\n",
    "df2 = results[-1].copy()\n",
    "df2['file'] = df2.genome_name.apply(lambda x: x.split('.')[0]+'_lcl.fasta')\n",
    "df2.fillna('-', inplace=True)\n",
    "df2['new predicted'] = df2.apply(lambda row: ':'.join([row['predicted_O'], row['predicted_H']]), axis=1)\n",
    "df2 = df2[['file', 'new predicted']]\n",
    "merged_df = df.merge(df2, on='file', how='left')\n",
    "merged_df.to_excel('modified_in_silico_summary.xlsx', index=False)\n",
    "merged_df[merged_df['new predicted'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_name</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>given_O</th>\n",
       "      <th>given_H</th>\n",
       "      <th>predicted_O</th>\n",
       "      <th>O_info</th>\n",
       "      <th>predicted_H</th>\n",
       "      <th>H_info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>GCA_001693635.1_ASM169363v1</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=O177:H21</td>\n",
       "      <td>O177</td>\n",
       "      <td>H21</td>\n",
       "      <td>O177</td>\n",
       "      <td>Alignment found</td>\n",
       "      <td>H21</td>\n",
       "      <td>Alignment found</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     genome_name     organism_name infraspecific_name given_O  \\\n",
       "444  GCA_001693635.1_ASM169363v1  Escherichia coli    strain=O177:H21    O177   \n",
       "\n",
       "    given_H predicted_O           O_info predicted_H           H_info  \n",
       "444     H21        O177  Alignment found         H21  Alignment found  "
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[-1][results[-1].loc[:, 'genome_name'].str.contains('001693635')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
